core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -680,6 +677,17 @@ private static void shiftMapping(Map<Integer, Integer> mapping, int startIndex,
 
   public Frame getInvoke(RelNode r, RelNode parent) {
     final Frame frame = dispatcher.invoke(r);
+    if (frame != null && parent instanceof Correlate && r instanceof Sort) {
```

Comment:
Does this code check only the parent or any ancestor? Is it valid to check only the parent?

Reply:
My understanding is that only when the sort is an input of Correlate, we can not decorrelate the sort's input (when the sort input has be decorrelated).

Reply:
Hmm, we should check all its ancestor, but i have no good way for this ? Should i put a flag there (indicating whether we are under a Correlate now ?) ?

Reply:
Finally i changes to check if the sort has a parent, if it does, that means the sort is definitely a branch of `Correlate` (if the sort input has been rewritten)

Reply:
Cool, thanks for checking!

Reply:
I think this discussion is relevant for https://issues.apache.org/jira/browse/CALCITE-4437